### ISSUE TITLE
Can't update `connectionDetails` for existing resources

### ISSUE THREAD

**Issue by jonnylangefeld** [crossplane/crossplane#3824]
During the investigation of #3823 I sometimes updated the `connectionDetails` of a resource in a composition for debugging and applied the new yaml to a cluster where the composition was already present before.

If I wanted to expose a second field for the service account key of the composition, like so
```
      connectionDetails:
        - fromConnectionSecretKey: private_key
          name: key.json
        - fromFieldPath: spec.forProvider.publicKeyType
          name: keyType
```
And added the new field `keyType` to the `connectionSecretKeys` of the XRD, I would expect to just update the composition and xrd and the new field should show up in the connection secret.

### What happened?
However, that is not the case. The connection secret that was present before remains unchanged.
Furthermore, if I delete the claim via `k delete -f claim.yaml` and recreate it, the connection secret gets created, but with no secret inside. Only after removing the XRD and the composition (which requires removal of all claims), will the new connection secret key be visible. 

This wouldn't work on large installations with many users.

### How can we reproduce it?
Use the following `XRD`, `Composition` and `Claim` resources:

<details>
<summary>`XRD` yaml:</summary>

```yaml
apiVersion: apiextensions.crossplane.io/v1
kind: CompositeResourceDefinition
metadata:
  name: xcruiseserviceaccounts.getcruise.com
spec:
  group: getcruise.com
  names:
    kind: XCruiseServiceAccount
    plural: xcruiseserviceaccounts
  claimNames:
    kind: CruiseServiceAccount
    plural: cruiseserviceaccounts
  connectionSecretKeys:
    - key.json
  versions:
    - name: v1alpha1
      served: true
      referenceable: true
      schema:
        openAPIV3Schema:
          type: object
          properties:
            spec: {}
```
</details>

<details>
<summary>`Composition` yaml:</summary>

```yaml
apiVersion: apiextensions.crossplane.io/v1
kind: Composition
metadata:
  name: xcruiseserviceaccounts.getcruise.com
  labels:
    provider: gcp
spec:
  publishConnectionDetailsWithStoreConfigRef:
    name: default
  compositeTypeRef:
    apiVersion: getcruise.com/v1alpha1
    kind: XCruiseServiceAccount
  resources:
    - name: serviceaccount
      base:
        # https://doc.crds.dev/github.com/upbound/provider-gcp/cloudplatform.gcp.upbound.io/ServiceAccount/v1beta1
        apiVersion: cloudplatform.gcp.upbound.io/v1beta1
        kind: ServiceAccount
        spec:
          forProvider: {}
      patches:
        - type: CombineFromComposite
          toFieldPath: spec.forProvider.description
          policy:
            fromFieldPath: Required
          combine:
            variables:
              - fromFieldPath: spec.claimRef.namespace
              - fromFieldPath: spec.claimRef.name
            strategy: string
            string:
              fmt: "%s/%s"
        - type: ToCompositeFieldPath
          fromFieldPath: status.atProvider.email
          toFieldPath: status.serviceAccountEmail
          policy:
            fromFieldPath: Required
    - name: serviceaccountkey
      base:
        # https://doc.crds.dev/github.com/upbound/provider-gcp/cloudplatform.gcp.upbound.io/ServiceAccountKey/v1beta1
        apiVersion: cloudplatform.gcp.upbound.io/v1beta1
        kind: ServiceAccountKey
        spec:
          forProvider:
            publicKeyType: TYPE_X509_PEM_FILE
            serviceAccountIdSelector:
              matchControllerRef: true
          publishConnectionDetailsTo:
            configRef:
              name: default
      patches:
        - type: CombineFromComposite
          combine:
            variables:
              - fromFieldPath: metadata.uid
            strategy: string
            string:
              fmt: "%s-serviceaccountkey"
          toFieldPath: spec.publishConnectionDetailsTo.name
          policy:
            fromFieldPath: Required
      connectionDetails:
        - fromConnectionSecretKey: private_key
          name: key.json
```
</details>

<details>
<summary>`Claim` yaml:</summary>

```yaml
apiVersion: getcruise.com/v1alpha1
kind: CruiseServiceAccount
metadata:
  name: jonnys-sa
spec:
  publishConnectionDetailsTo:
    configRef:
      name: default
    name: jonnys-sa
```
</details>

And apply them all. Then add another key to `connectionDetails` of the composition, like
```
        - fromFieldPath: spec.forProvider.publicKeyType
          name: keyType
```

And add that key to the XRD as well:
```
  connectionSecretKeys:
    - key.json
    - keyType
```

And apply all three resource again. The new key will not show up in the claim and the composite secret.

Now remove all three resources and apply the edited set, including the new key, again. In this case the `keyType` field will be present, as expected.

### What environment did it happen in?
Crossplane version: `v1.10.1`
GCP provider version: `v0.27.0`

**Comment by negz** [comment]
Yeah, this is a duplicate of https://github.com/crossplane/crossplane/issues/2024 which was closed for being stale. I'll reopen it and comment there.

**Comment by jonnylangefeld** [comment]
Ah sorry for the duplicate and thanks for explaining!
